### PR TITLE
Add arguments for API query parameters when fetching all Dandisets; support creating embargoed Dandisets

### LINKED ISSUE #1413: get_dandisets: Expose options (embargoed, empty, draft) of the /dandisets/ endpoint
I thought that as long as I authenticate I would also get embargoed datasets listed, as I believe is done in web UI.  But I found no Embargoed dandisets in returned list. Looking at our `get_dandisets` - apparently we do not expose any of the query parameters
image
for 
image

Immediate need is to get `embargoed` dandisets. I generally would not mind even adding "search" parameter.

### PR DESCRIPTION
**Pull request by jwodder**
Closes #1413.

### CODE REVIEW

File: dandi/dandiapi.py [hunk 1]
```diff
@@ -554,13 +554,62 @@ def get_dandiset(
...
+            "/dandisets/",
+            params={
+                "draft": draft,
+                "embargoed": embargoed,
+                "empty": empty,
+                "ordering": order,
+                "search": search,
+                "user": "me" if mine else None,
+            },
+        ):
```
Comment (yarikoptic) [inline dandi/dandiapi.py:612]: could we avoid specifying it at all if it was not specified in the call, i.e. `None`?

Comment (jwodder) [reply dandi/dandiapi.py:612]: @yarikoptic If you mean omitting the fields from `params`, that's unnecessary, as setting a param's value to `None` in `requests` is the same as not setting that param.

File: dandi/dandiapi.py [hunk 2]
```diff
@@ -554,13 +554,63 @@ def get_dandiset(
...
+        """
+        for data in self.paginate(
+            "/dandisets/",
+            params={
+                "draft": draft,
+                "embargoed": embargoed,
+                "empty": empty,
+                "ordering": order,
+                "search": search,
+                "user": "me" if mine else None,
```
Comment (yarikoptic) [inline dandi/dandiapi.py:611]: should we stick to API's interface here and expose as `user`?  May be admins (later?) might be able to query for some other users?

Comment (jwodder) [reply dandi/dandiapi.py:611]: No, I don't think having a parameter that can only take one value makes much sense.

**Review by yarikoptic — CHANGES_REQUESTED** [review]
Let's add a test or two. 

ATM we do not really test/deal much with embargoed dandisets seems to me, primarily since we do not even expose `embargo` option:

```
dandi/dandiapi.py:    def create_dandiset(self, name: str, metadata: dict[str, Any]) -> RemoteDandiset:
```

So let's as part of this PR expose that option, create also embargoed dandiset and test listing here for a set of query parameters to see that they map correctly to the API call (as you mentioned -- requests would simply ignore if we do not change them from `None`)

File: dandi/dandiapi.py (hunk 2 above)
Comment (yarikoptic) [reply dandi/dandiapi.py:611]: I agree... as we are to RF order vs ordering, please file another issue on this particular one. May be indeed the intent would be to expose querying for other users as well. If not -- we better recommend to refactor it to `mine`.

Comment (jwodder) [reply dandi/dandiapi.py:611]: Issue filed: https://github.com/dandi/dandi-archive/issues/1879

**Comment by jwodder** [comment]
@yarikoptic 

> test listing here for a set of query parameters to see that they map correctly to the API call

What exactly should be tested?  Should there just be a test for each `bool` argument that checks that a matching Dandiset is/isn't returned when the argument is `true`/`false`?

Note that there can't be a test that checks the entirety of the `get_dandisets()` return value, as the test Dandisets created for most tests persist in later tests, and it's possible for tests to be skipped or run out of order, so there's no guarantee about what Dandisets will exist in the test Archive when a test runs.

**Comment by yarikoptic** [comment]
> Note that there can't be a test that checks the entirety of the `get_dandisets()` return value, as the test Dandisets created for most tests persist in later tests, and it's possible for tests to be skipped or run out of order, so there's no guarantee about what Dandisets will exist in the test Archive when a test runs.

I didn't realize that -- I thought that we just have a fixture to populate a number of dandisets and then test on them.


> What exactly should be tested? Should there just be a test for each `bool` argument that checks that a matching Dandiset is/isn't returned when the argument is `true`/`false`?

I think testing for that would suffice. Fixture could ensure to populate at least one dandiset to satisfy each criteria, then (with the light of above) - we could test that there is at least one dandiset satisfying the desired criteria.

**Comment by yarikoptic** [comment]
Thank you!

**Pull request merged**